From: Borislav Petkov <bp@suse.de>
Date: Wed, 10 Jan 2018 00:10:42 +0100
Subject: x86/CPU: Sync CPU feature flags late
Patch-mainline: not yet, for Teradata testing
References: bsc#1075994 bsc#1075091

This is for the case where we need to set feature flags late, like, for
example, after late microcode patch has been loaded which has enabled
new CPUID bits.

This has no effect on alternatives patching.

Signed-off-by: Borislav Petkov <bp@suse.de>
Cc: Peter Zijlstra <peterz@infradead.org>
Cc: x86@kernel.org
---
 arch/x86/kernel/cpu/common.c |   19 +++++++++++++++++++
 arch/x86/kernel/cpu/cpu.h    |    3 +++
 2 files changed, 22 insertions(+)

--- a/arch/x86/kernel/cpu/common.c
+++ b/arch/x86/kernel/cpu/common.c
@@ -647,6 +647,25 @@ static void apply_forced_caps(struct cpu
 	}
 }
 
+/*
+ * This late synchronization of CPU caps has no effect on alternatives patching
+ * but updates the visible feature bits per CPU.
+ *
+ * Callers need to disable CPU hotplug around it.
+ */
+void cpu_caps_sync_late(void)
+{
+	int cpu;
+
+
+	for_each_online_cpu(cpu) {
+		struct cpuinfo_x86 *c = &cpu_data(cpu);
+
+		apply_forced_caps(c);
+	}
+}
+EXPORT_SYMBOL_GPL(cpu_caps_sync_late);
+
 void __cpuinit get_cpu_cap(struct cpuinfo_x86 *c)
 {
 	u32 tfms, xlvl;
--- a/arch/x86/kernel/cpu/cpu.h
+++ b/arch/x86/kernel/cpu/cpu.h
@@ -34,6 +34,9 @@ extern const struct cpu_dev *const __x86
 			    *const __x86_cpu_dev_end[];
 
 extern void get_cpu_cap(struct cpuinfo_x86 *c);
+
+extern void cpu_caps_sync_late(void);
+
 extern void cpu_detect_cache_sizes(struct cpuinfo_x86 *c);
 extern void get_cpu_cap(struct cpuinfo_x86 *c);
 
